feat!: Add #[non_exhaustive] to ChangeSet structs

notmandatory opened the pull request:
### Description

Add #[non_exhaustive] to ChangeSet structs in tx_graph, keychain_txout, local_chain.

Add #[non_exhaustive] attribute to the ChangeSet structs to prevent downstream crates from exhaustively matching or constructing instances using struct literal syntax, allowing future additions without breaking compatibility.

The change improves forward compatibility by allowing new fields to be added without breaking downstream code in future releases.

### Notes to the reviewers

The motivation for this is to prevent breaking changes in these structs but will still require downstream persistence crates to implement and support any new fields added. 

@ValuedMammal suggested creating a persistence test suite which I think is a great idea but should be done in a separate PR.

### Changelog notice

#### Breaking Changes
- Added `#[non_exhaustive]` attribute to `ChangeSet` structs in tx_graph, keychain_txout, local_chain modules.
  - Match expressions on these structs must now include wildcard (`..`) patterns.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo +nightly fmt` and `cargo clippy` before committing

#### New Features:

* [ ] I've added tests for the new feature
* [ ] I've added docs for the new feature

#### Bugfixes:

* [ ] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [ ] I'm linking the issue being fixed by this PR

evanlinjin commented:
The persistence test suite will make this work. It's unfortunate we didn't think about that solution earlier.

notmandatory committed:
feat!: Add #[non_exhaustive] to ChangeSet structs in tx_graph, keychain_txout, local_chain

Add #[non_exhaustive] attribute to the ChangeSet structs to prevent downstream
crates from exhaustively matching or constructing instances using struct
literal syntax, allowing future additions without breaking compatibility.

The change improves forward compatibility by allowing new fields to be added
without breaking downstream code in future releases.

notmandatory commented:
I fixed the tests, I don't think they look too bad with the non-exhaustive change sets.

tnull reviewed (commented):
The `wallet::ChangeSet` struct used to be `non_exhaustive`, which was reverted after I had argued in https://github.com/bitcoindevkit/bdk/issues/1591 that it shouldn't be. 

Why do we now think the sub-changesets should be `non_exhaustive`? To restate my point: having your code break is a good thing if you're handling the changeset types manually, because then you're forced to update your persistence. If you force a non-breaking API via `non_exhaustive`, users could miss persisting important fields after they had been added.

notmandatory commented:
Ah ok I completely forgot we discussed this last year. It seems we have conflicting goals:
* I want to be able to add new, default-able fields in persisted data as non-breaking changes so we don't need to do a new major release every time we add a feature that requires persisting new data. 
* You want existing persistence code to break when new fields are added to persisted data so all implementations are forced to stay up-to-date. 

I think a part of the problem is we're trying to support a persistence API that is tightly coupled to our ChangeSet structs. I feel like there should be a way to support both goals. Possibly with some sort of persistence store traits or persistence conformance test suite, maybe both. But I'll do some more thinking on it, and open to any other ideas.

tnull commented:
> * I want to be able to add new, default-able fields in persisted data as non-breaking changes so we don't need to do a new major release every time we add a feature that requires persisting new data.

`ChangeSet` in part of the public API. If you add new data fields that a user is required to persist, it's by definition a breaking change, no?

>  Possibly with some sort of persistence store traits or persistence conformance test suite

What you describe here is a BDK-internal data model/serialization logic, which was decided against some time ago. But yeah, if that is the conclusion, I'd be very happy about that outcome (especially if it considers forwards compat.). ;)

sdmg15 commented on crates/chain/tests/test_tx_graph.rs:
```diff
@@ -1470,39 +1451,27 @@ fn tx_graph_update_conversion() {
-        }
-    );
+    let mut expected_changeset = ChangeSet::default();
+    expected_changeset.last_seen = [(hash!("tx1"), 1000001)].into();
```
Nit: Since we're formatting the number, maybe we could do that too here.

sdmg15 commented on crates/chain/src/indexer/keychain_txout.rs:
```diff
@@ -1047,6 +1047,7 @@ impl<K: core::fmt::Display + core::fmt::Debug> core::error::Error for InsertDesc
 #[derive(Clone, Debug, Default, PartialEq)]
 #[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
 #[must_use]
+#[non_exhaustive]
```
Will it be interesting to either through doc comment or inside the book to add about the rational here?

sdmg15 reviewed (commented):
tAck.

`cargo test --all-targets`

<img width="704" height="178" alt="Image" src="https://github.com/user-attachments/assets/b3ee8b8a-7158-450b-be5e-a8b46fd760c7" />

oleonardolima reviewed (commented):
cACK dd97143463111b49126b4e6100f5b80cb323df11

evanlinjin commented:
After a discussion with @oleonardolima and @ValuedMammal, we concluded that there is no way to introduce this feature without introducing dire footguns for the caller (even if we have #2012 in place).

Image the following situation:

* Latest `bdk_chain` version is `0.25.0`.
* Caller is using a persistence crate `bdk_mongodb = 0.1` which uses persistence tests from `bdk_chain = 0.25.0`.
* Caller has `bdk_chain = 0.25` in their Cargo (patch version will be bumped automatically).
* We introduce new changeset fields in `bdk_chain = 0.25.1` and release it.
* Maintainer of `bdk_mongodb` does not update persistence tests to use the version in `bdk_chain = 0.25.1`.
* Caller does `cargo update` in their project and changesets get new fields. `bdk_mongodb` ignores those new fields - potentially loosing data.

The benefits of this does not outweigh this massive footgun. We rarely introduce new changeset fields anyway.

notmandatory commented:
Thanks for the detailed thought on this. I was worried we'd have too many changes on this struct but in practice it seems to be going OK with us making breaking version bump when we need to add fields to it. We should still try to limit changes as much as possible.

tnull commented:
> * We introduce new changeset fields in `bdk_chain = 0.25.1` and release it.

Hmm, well, isn't `ChangeSet` part of the public API, so you can't/shouldn't change it on a patch release anyways?
